Mathlib/Order/WellQuasiOrder.lean

```diff
@@ -0,0 +1,151 @@
+with no infinite antichains.
+
+## Main definitions
+
+* `WellQuasiOrdered`: a predicate for WQO unbundled relations
+* `WellQuasiOrderedLE`: a typeclass for a bundled WQO `≤` relation
+
+## TODO
+
+* Define `Set.PartiallyWellOrderedOn` and `Set.IsPWO` in terms of these predicates, and rename them
+  to match.
+-/
+
+variable {α β : Type*} {r : α → α → Prop} {s : β → β → Prop}
+
+/-- A well quasi-order or WQO is a relation such that any infinite sequence contains an infinite
+monotonic subsequence, or equivalently, two elements `f m` and `f n` with `m < n` and
+`r (f m) (f n)`.
+
+For a preorder, this is equivalent to having a well-founded order with no infinite antichains.
+
+Despite the nomenclature, we don't require the relation to be preordered. Moreover, a well
+quasi-order will not in general be a well-order. -/
+def WellQuasiOrdered (r : α → α → Prop) : Prop :=
+  ∀ f : ℕ → α, ∃ m n : ℕ, m < n ∧ r (f m) (f n)
+
+theorem wellQuasiOrdered_of_isEmpty [IsEmpty α] (r : α → α → Prop) : WellQuasiOrdered r :=
+  fun f ↦ isEmptyElim (f 0)
+
+theorem IsAntichain.finite_of_wellQuasiOrdered {s : Set α} (hs : IsAntichain r s)
+    (hr : WellQuasiOrdered r) : s.Finite := by
+  refine Set.not_infinite.1 fun hi => ?_
+  obtain ⟨m, n, hmn, h⟩ := hr fun n => hi.natEmbedding _ n
+  exact hmn.ne ((hi.natEmbedding _).injective <| Subtype.val_injective <|
+    hs.eq (hi.natEmbedding _ m).2 (hi.natEmbedding _ n).2 h)
+
+theorem Finite.wellQuasiOrdered (r : α → α → Prop) [Finite α] [IsRefl α r] :
+    WellQuasiOrdered r := by
+  intro f
+  obtain ⟨m, n, h, hf⟩ := Set.finite_univ.exists_lt_map_eq_of_forall_mem (f := f)
+    fun _ ↦ Set.mem_univ _
+  exact ⟨m, n, h, hf ▸ refl _⟩
+
+theorem WellQuasiOrdered.exists_monotone_subseq [IsPreorder α r] (h : WellQuasiOrdered r)
+    (f : ℕ → α) : ∃ g : ℕ ↪o ℕ, ∀ m n, m ≤ n → r (f (g m)) (f (g n)) := by
+  obtain ⟨g, h1 | h2⟩ := exists_increasing_or_nonincreasing_subseq r f
+  · refine ⟨g, fun m n hle => ?_⟩
+    obtain hlt | rfl := hle.lt_or_eq
+    exacts [h1 m n hlt, refl_of r _]
+  · obtain ⟨m, n, hlt, hle⟩ := h (f ∘ g)
+    cases h2 m n hlt hle
+
+theorem wellQuasiOrdered_iff_exists_monotone_subseq [IsPreorder α r] :
+    WellQuasiOrdered r ↔ ∀ f : ℕ → α, ∃ g : ℕ ↪o ℕ, ∀ m n : ℕ, m ≤ n → r (f (g m)) (f (g n)) := by
+  constructor <;> intro h f
+  · exact h.exists_monotone_subseq f
+  · obtain ⟨g, gmon⟩ := h f
+    exact ⟨_, _, g.strictMono Nat.zero_lt_one, gmon _ _ (Nat.zero_le 1)⟩
+
+theorem WellQuasiOrdered.prod [IsPreorder α r] (hr : WellQuasiOrdered r) (hs : WellQuasiOrdered s) :
+    WellQuasiOrdered fun a b : α × β ↦ r a.1 b.1 ∧ s a.2 b.2 := by
+  intro f
+  obtain ⟨g, h₁⟩ := hr.exists_monotone_subseq (Prod.fst ∘ f)
+  obtain ⟨m, n, h, hf⟩ := hs (Prod.snd ∘ f ∘ g)
+  exact ⟨g m, g n, g.strictMono h, h₁ _ _ h.le, hf⟩
+
+/-- A typeclass for an order with a well quasi-ordered `≤` relation.
+
+Note that this is unlike `WellFoundedLT`, which instead takes a `<` relation. -/
+@[mk_iff wellQuasiOrderedLE_iff']
```

Comment:
I'd like to avoid adding more primed lemmas to mathlib. Either `wellQuasiOrderedLE_def` or `wellQuasiOrderedLE_iff_wellQuasiOrdered` would be better here.